{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from random import randrange \n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import argparse\n",
    "import subprocess\n",
    "from time import clock\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import groupby\n",
    "import itertools\n",
    "from Bio.Alphabet import generic_dna\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Alphabet import IUPAC, Gapped\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "from Bio import Phylo\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "from sys import argv\n",
    "from Bio import AlignIO\n",
    "from time import time\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import seaborn as sns; sns.set()\n",
    "from scipy import stats\n",
    "import matplotlib.backends.backend_pdf\n",
    "import pylab as pl"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "root = \"/home/pavel/Desktop/Work/Project/All_data/Worms/\"\n",
    "files_txt = sorted([f for f in os.listdir(root) if os.path.splitext(f)[1] == '.fasta'])\n",
    "\n",
    "os.chdir(root)\n",
    "for i in files_txt:\n",
    "    tmp = \".\".join(i.split(\".\")[:-1])\n",
    "    print(tmp)\n",
    "    try:\n",
    "        line = 'clustalw -INFILE={in_file} -OUTFILE={out_file} -TYPE=DNA -OUTPUT=FASTA'.format(in_file = i, \n",
    "                                                                                              out_file = tmp + \"_al.fasta\")\n",
    "        p = subprocess.Popen(line, shell=True)\n",
    "        p.wait()\n",
    "        \n",
    "        line1 = 'snp-sites -mvp -o out.fasta {in_file}'.format(in_file = root + tmp + \"_al.fasta\")\n",
    "        p = subprocess.Popen(line1, shell=True)\n",
    "        p.wait()\n",
    "                \n",
    "        line2 = 'vcftools --hap-r2 --vcf out.fasta.vcf  --out out_vcftools'\n",
    "        p = subprocess.Popen(line2, shell=True)\n",
    "        p.wait()\n",
    "           \n",
    "        df = pd.read_csv(root + \"out_vcftools.hap.ld\", sep='\\t',index_col=False)\n",
    "        df[\"Len\"] = df[\"POS2\"] - df[\"POS1\"]\n",
    "    \n",
    "        fig = sns.relplot(x=\"Len\", y=\"R^2\", kind=\"line\", data=df)\n",
    "        plt.gcf().set_size_inches(25, 15)\n",
    "        axes = fig.axes\n",
    "        axes[0,0].set_ylim(0,1.01)\n",
    "        plt.show()\n",
    "        \n",
    "        !rm out.fasta.vcf\n",
    "        !rm out_vcftools.hap.ld\n",
    "        !rm out_vcftools.log\n",
    "        !rm out.fasta.phylip\n",
    "        !rm out.fasta.snp_sites.aln\n",
    "        \n",
    "    except FileNotFoundError:\n",
    "        continue\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rm: невозможно удалить '*_al.*': Нет такого файла или каталога\n",
      "Arenicola_marina\n",
      "Bivalvia_control.fasta.aligned\n",
      "Laonice_cirrata\n",
      "Laonice_cirrata_corr\n",
      "Marenzelleria_arctica\n",
      "Priapulus_caudatus\n",
      "Rhynchospio_glutaea\n",
      "Scolopus_from_Alyona\n",
      "Spiophanes_bombyx_4\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 432x288 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 864x576 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 864x576 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 864x576 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 864x576 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 864x576 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 864x576 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 864x576 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 864x576 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 864x576 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "import matplotlib.backends.backend_pdf\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "import pylab as pl\n",
    "\n",
    "root = \"/home/pavel/Desktop/Work/Project/All_data/Worms/\"\n",
    "files_txt = sorted([f for f in os.listdir(root) if os.path.splitext(f)[1] == '.fasta'])\n",
    "\n",
    "out_pdf = r'/home/pavel/Desktop/Work/Project/All_data/Worms.pdf'\n",
    "pdf = matplotlib.backends.backend_pdf.PdfPages(out_pdf)\n",
    "\n",
    "\n",
    "def mva(a, window):\n",
    "    arr = []\n",
    "    l = 0\n",
    "    r = window\n",
    "    for i in range(1, window + 1):\n",
    "        #print(i)\n",
    "        r = i * window\n",
    "        s = a[l:r]\n",
    "        #print(s)\n",
    "        l = r\n",
    "        arr.append(np.mean(s))\n",
    "    return(arr)\n",
    "\n",
    "\n",
    "figs = plt.figure()\n",
    "os.chdir(root)\n",
    "\n",
    "\n",
    "p = subprocess.Popen(\"rm *_al.*\", shell=True)\n",
    "p.wait()\n",
    "\n",
    "for i in files_txt:\n",
    "    tmp = \".\".join(i.split(\".\")[:-1])\n",
    "    print(tmp)\n",
    "    try:\n",
    "        line = 'clustalw -INFILE={in_file} -OUTFILE={out_file} -TYPE=DNA -OUTPUT=FASTA'.format(in_file = i, \n",
    "                                                                                              out_file = tmp + \"_al.fasta\")\n",
    "        p = subprocess.Popen(line, shell=True)\n",
    "        p.wait()\n",
    "        \n",
    "        line1 = 'snp-sites -mvp -o out.fasta {in_file}'.format(in_file = root + tmp + \"_al.fasta\")\n",
    "        p = subprocess.Popen(line1, shell=True)\n",
    "        p.wait()\n",
    "                \n",
    "        line2 = 'vcftools --hap-r2 --vcf out.fasta.vcf  --out out_vcftools'\n",
    "        p = subprocess.Popen(line2, shell=True)\n",
    "        p.wait()\n",
    "           \n",
    "        df = pd.read_csv(root + \"out_vcftools.hap.ld\", sep='\\t',index_col=False)\n",
    "        df[\"Len\"] = df[\"POS2\"] - df[\"POS1\"]\n",
    "    \n",
    "        df = df.sort_values(by=['Len'])\n",
    "        x = df[\"Len\"]\n",
    "        y = df[\"R^2\"]\n",
    "    \n",
    "        window = len(x)//5\n",
    "        a = mva(x, window)\n",
    "        b = mva(y, window)\n",
    "        \n",
    "        f, ax = plt.subplots(figsize=(12, 8))\n",
    "        #fig = sns.scatterplot(x=\"Len\", y=\"R^2\", data=df, ax=ax, s=80, marker=\"o\")\n",
    "        fig = sns.regplot(x=df[\"Len\"], y=df[\"R^2\"], fit_reg=False, scatter_kws={\"color\":\"blue\",\"alpha\":0.1,\"s\":20} )\n",
    "        fig = sns.lineplot(x=\"x\", y=\"y\", data=pd.DataFrame(data={\"x\":a, \"y\":b}), ax=ax, color=\"red\")\n",
    "        \n",
    "        plt.title('R^2 plot. ' + tmp)\n",
    "        plt.xlabel(\"Len\")\n",
    "        plt.ylabel(\"R^2\")\n",
    "        \n",
    "        fig.set(ylim=(-0.05, 1.05))\n",
    "        pdf.savefig()\n",
    "        plt.clf()\n",
    "        \n",
    "        !rm out.fasta.vcf\n",
    "        !rm out_vcftools.hap.ld\n",
    "        !rm out_vcftools.log\n",
    "        !rm out.fasta.phylip\n",
    "        !rm out.fasta.snp_sites.aln\n",
    "        \n",
    "    except FileNotFoundError:\n",
    "        continue\n",
    "    #break\n",
    "        \n",
    "pdf.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 432x288 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "import matplotlib.backends.backend_pdf\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "import pylab as pl\n",
    "\n",
    "root = \"/home/pavel/Desktop/Work/Project/All_data/Sach/\"\n",
    "files_txt = sorted([f for f in os.listdir(root) if f.split(\".\")[-2:] == ['aligned', 'fasta']])\n",
    "\n",
    "out_pdf = r'/home/pavel/Desktop/Work/Project/All_data/Sach.pdf'\n",
    "pdf = matplotlib.backends.backend_pdf.PdfPages(out_pdf)\n",
    "\n",
    "\n",
    "def mva(a, window):\n",
    "    arr = []\n",
    "    l = 0\n",
    "    r = window\n",
    "    for i in range(1, window + 1):\n",
    "        #print(i)\n",
    "        r = i * window\n",
    "        s = a[l:r]\n",
    "        #print(s)\n",
    "        l = r\n",
    "        arr.append(np.mean(s))\n",
    "    return(arr)\n",
    "\n",
    "\n",
    "figs = plt.figure()\n",
    "os.chdir(root)\n",
    "for i in files_txt:\n",
    "    tmp = \".\".join(i.split(\".\")[:-1])\n",
    "    print(tmp)\n",
    "    try:\n",
    "\n",
    "        line1 = 'snp-sites -mvp -o out.fasta {in_file}'.format(in_file = root + i)\n",
    "        p = subprocess.Popen(line1, shell=True)\n",
    "        p.wait()\n",
    "                \n",
    "        line2 = 'vcftools --hap-r2 --vcf out.fasta.vcf  --out out_vcftools'\n",
    "        p = subprocess.Popen(line2, shell=True)\n",
    "        p.wait()\n",
    "           \n",
    "        df = pd.read_csv(root + \"out_vcftools.hap.ld\", sep='\\t',index_col=False)\n",
    "        df[\"Len\"] = df[\"POS2\"] - df[\"POS1\"]\n",
    "    \n",
    "        df = df.sort_values(by=['Len'])\n",
    "        x = df[\"Len\"]\n",
    "        y = df[\"R^2\"]\n",
    "    \n",
    "        window = len(x)//5\n",
    "        a = mva(x, window)\n",
    "        b = mva(y, window)\n",
    "        \n",
    "        f, ax = plt.subplots(figsize=(12, 8))\n",
    "        #fig = sns.scatterplot(x=\"Len\", y=\"R^2\", data=df, ax=ax, s=80, marker=\"o\")\n",
    "        fig = sns.regplot(x=df[\"Len\"], y=df[\"R^2\"], fit_reg=False, scatter_kws={\"color\":\"blue\",\"alpha\":0.1,\"s\":20} )\n",
    "        fig = sns.lineplot(x=\"x\", y=\"y\", data=pd.DataFrame(data={\"x\":a, \"y\":b}), ax=ax, color=\"red\")\n",
    "        \n",
    "        plt.title('R^2 plot. ' + tmp)\n",
    "        plt.xlabel(\"Len\")\n",
    "        plt.ylabel(\"R^2\")\n",
    "        \n",
    "        fig.set(ylim=(-0.05, 1.05))\n",
    "        pdf.savefig()\n",
    "        plt.clf()\n",
    "        \n",
    "        !rm out.fasta.vcf\n",
    "        !rm out_vcftools.hap.ld\n",
    "        !rm out_vcftools.log\n",
    "        !rm out.fasta.phylip\n",
    "        !rm out.fasta.snp_sites.aln\n",
    "        plt.close('all')\n",
    "    except FileNotFoundError:\n",
    "        continue\n",
    "    #break\n",
    "        \n",
    "pdf.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saccharomyces_Cerevisiae_complete_genomes.fasta.aligned\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import matplotlib.backends.backend_pdf\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "import pylab as pl\n",
    "\n",
    "circle_mode = True\n",
    "aligned = True\n",
    "name_format = \"aligned\"\n",
    "\n",
    "root = \"/home/pavel/Desktop/Work/Project/All_data/\"\n",
    "\n",
    "\n",
    "if name_format == \"aligned_al\":\n",
    "    files_txt = sorted([f for f in os.listdir(root) if f.split(\".\")[-2:] == ['aligned_al', 'fasta']])\n",
    "elif name_format == \"aligned\":\n",
    "    files_txt = sorted([f for f in os.listdir(root) if f.split(\".\")[-2:] == ['aligned', 'fasta']])\n",
    "else:\n",
    "    files_txt = sorted([f for f in os.listdir(root) if f.split(\".\")[-1:] == ['fasta']])\n",
    "out_pdf = r'/home/pavel/Desktop/Work/Project/All_data/Saccharomyces.pdf'\n",
    "pdf = matplotlib.backends.backend_pdf.PdfPages(out_pdf)\n",
    "\n",
    "\n",
    "def mva(a, window):\n",
    "    arr = []\n",
    "    l = 0\n",
    "    r = window\n",
    "    for i in range(1, window + 1):\n",
    "        #print(i)\n",
    "        r = i * window\n",
    "        s = a[l:r]\n",
    "        #print(s)\n",
    "        l = r\n",
    "        arr.append(np.mean(s))\n",
    "    return(arr)\n",
    " \n",
    "\n",
    "figs = plt.figure()\n",
    "os.chdir(root)\n",
    "for i in files_txt:\n",
    "    \n",
    "    with open(root + i, \"r\") as f:\n",
    "        seqs={}\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            if len(line) > 0:\n",
    "                if line[0] == \">\":\n",
    "                    words = line.split()\n",
    "                    name = words[0][1:]\n",
    "                    seqs[name] = ''\n",
    "                else:\n",
    "                    seqs[name] = seqs[name] + line\n",
    "        seq_len = len(list(seqs.values())[0])\n",
    "    \n",
    "    tmp = \".\".join(i.split(\".\")[:-1])\n",
    "    print(tmp)\n",
    "    try:\n",
    "\n",
    "        if aligned == False:\n",
    "            line = 'clustalw -INFILE={in_file} -OUTFILE={out_file} -TYPE=DNA -OUTPUT=FASTA'.format(in_file = i, \n",
    "                                                                                              out_file = tmp + \"_al.fasta\")\n",
    "            p = subprocess.Popen(line, shell=True)\n",
    "            p.wait()\n",
    "        \n",
    "        \n",
    "        line1 = 'snp-sites -mvp -o out.fasta {in_file}'.format(in_file = root + i)\n",
    "        p = subprocess.Popen(line1, shell=True)\n",
    "        p.wait()\n",
    "                \n",
    "        line2 = 'vcftools --hap-r2 --vcf out.fasta.vcf  --out out_vcftools'\n",
    "        p = subprocess.Popen(line2, shell=True)\n",
    "        p.wait()\n",
    "           \n",
    "        df = pd.read_csv(root + \"out_vcftools.hap.ld\", sep='\\t',index_col=False)\n",
    "        if len(list(df[\"POS1\"])) > 100000:\n",
    "            df1 = df.sample(100000)\n",
    "        else:\n",
    "            df1 = df\n",
    "        if circle_mode == True:\n",
    "            POS1 = list(df1[\"POS1\"])\n",
    "            POS2 = list(df1[\"POS2\"])\n",
    "            #print(POS1[:100])\n",
    "            #print(POS2[:100])\n",
    "            OUT = [None]*len(POS1)\n",
    "            for line in range(len(POS1)):\n",
    "                #print(line)\n",
    "                #print(min(POS2[line] - POS1[line], seq_len - POS2[line] - POS1[line]))\n",
    "                #print(POS2[line] - POS1[line])\n",
    "                OUT[line] = min(abs(POS2[line] - POS1[line]), seq_len - abs(POS2[line] - POS1[line]))\n",
    "            df1[\"Len\"] = OUT\n",
    "        else:\n",
    "            df1[\"Len\"] = df1[\"POS2\"] - df1[\"POS1\"]\n",
    "    \n",
    "        df1 = df1.sort_values(by=['Len'])\n",
    "        x = df1[\"Len\"]\n",
    "        y = df1[\"R^2\"]\n",
    "    \n",
    "        window = len(x)//100\n",
    "        a = mva(x, window)\n",
    "        b = mva(y, window)\n",
    "        \n",
    "        f, ax = plt.subplots(figsize=(12, 8))\n",
    "        #fig = sns.scatterplot(x=\"Len\", y=\"R^2\", data=df, ax=ax, s=80, marker=\"o\")\n",
    "        fig = sns.regplot(x=df1[\"Len\"], y=df1[\"R^2\"], fit_reg=False, scatter_kws={\"color\":\"blue\",\"alpha\":0.1,\"s\":20} )\n",
    "        fig = sns.lineplot(x=\"x\", y=\"y\", data=pd.DataFrame(data={\"x\":a, \"y\":b}), ax=ax, color=\"red\")\n",
    "        \n",
    "        plt.title('R^2 plot. ' + tmp)\n",
    "        plt.xlabel(\"Len\")\n",
    "        plt.ylabel(\"R^2\")\n",
    "        \n",
    "        fig.set(ylim=(-0.05, 1.05))\n",
    "        pdf.savefig()\n",
    "        plt.clf()\n",
    "        \n",
    "        !rm out.fasta.vcf\n",
    "        !rm out_vcftools.hap.ld\n",
    "        !rm out_vcftools.log\n",
    "        !rm out.fasta.phylip\n",
    "        !rm out.fasta.snp_sites.aln\n",
    "        plt.close('all')\n",
    "    except FileNotFoundError:\n",
    "        continue\n",
    "    print(\"done\")\n",
    "        \n",
    "pdf.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import matplotlib.backends.backend_pdf\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "import pylab as pl\n",
    "\n",
    "circle_mode = True\n",
    "aligned = True\n",
    "\n",
    "root = \"/home/pavel/Desktop/Work/Project/Simulation/result_4/\"\n",
    "\n",
    "\n",
    "files_txt = sorted([f for f in os.listdir(root) if f.split(\".\")[-1:] == ['csv']])\n",
    "\n",
    "LD_df_1500_300_0.0_0_4_1000.csv\n",
    "\n",
    "out_pdf = r'/home/pavel/Desktop/Work/Project/Simulation/LD_tmp.pdf'\n",
    "pdf = matplotlib.backends.backend_pdf.PdfPages(out_pdf)\n",
    "\n",
    "\n",
    "def mva(a, window):\n",
    "    arr = []\n",
    "    l = 0\n",
    "    r = window\n",
    "    for i in range(1, window + 1):\n",
    "        #print(i)\n",
    "        r = i * window\n",
    "        s = a[l:r]\n",
    "        #print(s)\n",
    "        l = r\n",
    "        arr.append(np.mean(s))\n",
    "    return(arr)\n",
    " \n",
    "\n",
    "figs = plt.figure()\n",
    "os.chdir(root)\n",
    "for i in files_txt:\n",
    "\n",
    "    if len(list(df[\"POS1\"])) > 100000:\n",
    "            df1 = df.sample(100000)\n",
    "        else:\n",
    "            df1 = df\n",
    "        if circle_mode == True:\n",
    "            POS1 = list(df1[\"POS1\"])\n",
    "            POS2 = list(df1[\"POS2\"])\n",
    "            #print(POS1[:100])\n",
    "            #print(POS2[:100])\n",
    "            OUT = [None]*len(POS1)\n",
    "            for line in range(len(POS1)):\n",
    "                #print(line)\n",
    "                #print(min(POS2[line] - POS1[line], seq_len - POS2[line] - POS1[line]))\n",
    "                #print(POS2[line] - POS1[line])\n",
    "                OUT[line] = min(abs(POS2[line] - POS1[line]), seq_len - abs(POS2[line] - POS1[line]))\n",
    "            df1[\"Len\"] = OUT\n",
    "        else:\n",
    "            df1[\"Len\"] = df1[\"POS2\"] - df1[\"POS1\"]\n",
    "    \n",
    "        df1 = df1.sort_values(by=['Len'])\n",
    "        x = df1[\"Len\"]\n",
    "        y = df1[\"R^2\"]\n",
    "    \n",
    "        window = len(x)//100\n",
    "        a = mva(x, window)\n",
    "        b = mva(y, window)\n",
    "        \n",
    "        f, ax = plt.subplots(figsize=(12, 8))\n",
    "        #fig = sns.scatterplot(x=\"Len\", y=\"R^2\", data=df, ax=ax, s=80, marker=\"o\")\n",
    "        fig = sns.regplot(x=df1[\"Len\"], y=df1[\"R^2\"], fit_reg=False, scatter_kws={\"color\":\"blue\",\"alpha\":0.1,\"s\":20} )\n",
    "        fig = sns.lineplot(x=\"x\", y=\"y\", data=pd.DataFrame(data={\"x\":a, \"y\":b}), ax=ax, color=\"red\")\n",
    "        \n",
    "        plt.title('R^2 plot. ' + tmp)\n",
    "        plt.xlabel(\"Len\")\n",
    "        plt.ylabel(\"R^2\")\n",
    "        \n",
    "        fig.set(ylim=(-0.05, 1.05))\n",
    "        pdf.savefig()\n",
    "        plt.clf()\n",
    "        \n",
    "        !rm out.fasta.vcf\n",
    "        !rm out_vcftools.hap.ld\n",
    "        !rm out_vcftools.log\n",
    "        !rm out.fasta.phylip\n",
    "        !rm out.fasta.snp_sites.aln\n",
    "        plt.close('all')\n",
    "    except FileNotFoundError:\n",
    "        continue\n",
    "    print(\"done\")\n",
    "        \n",
    "pdf.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "~/Desktop/Work/Project/trimAl/source/trimal -in ./Saccharomyces_Cerevisiae_complete_genomes.fasta.aligned.fasta -out ./out.fasta -automated1; ~/Desktop/Work/Project/trimAl/source/trimal -in ./out.fasta -out ./out2.fasta -nogaps\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
